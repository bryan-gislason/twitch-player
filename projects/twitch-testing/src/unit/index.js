import { ThemeProvider } from "@twitch-player/themes/dist/context";
import { light as lightTheme } from "@twitch-player/themes/dist/themes";
import { OverridesProvider } from "@twitch-player/ui/dist/context";
import React from "react";
import { act } from "react-dom/test-utils";
import { MemoryRouter as Router } from "react-router";
import { fireEvent, render } from "react-testing-library";

const customRender = (node, ...options) => {
  return render(
    <ThemeProvider theme={lightTheme}>
      <OverridesProvider>
        <Router>{node}</Router>
      </OverridesProvider>
    </ThemeProvider>,
    ...options
  );
};

fireEvent["press"] = (node, init) => {
  fireEvent.mouseDown(node, init);
  fireEvent.mouseUp(node, init);
};

fireEvent["resize"] = (node, init) => {
  const { target: { left = 0, top = 0, width = 0, height = 0 } = {} } = init;

  Object.defineProperties(window.HTMLElement.prototype, {
    offsetLeft: { get: () => left },
    offsetTop: { get: () => top },
    offsetWidth: { get: () => width },
    offsetHeight: { get: () => height },
  });

  const eventInit = { bubbles: false, cancelable: false };
  const event = new Event(node, eventInit);

  act(() => {
    fireEvent(node, event);
    // eslint-disable-next-line no-undef
    jest.runAllTimers();
  });
};

export * from "react-testing-library";

export { customRender as render };
