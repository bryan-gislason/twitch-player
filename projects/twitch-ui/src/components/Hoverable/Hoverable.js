import PropTypes from "prop-types";
import { Children, Component, cloneElement } from "react";
import { isHoverEnabled } from "../../utils";

class Hoverable extends Component {
  constructor(props) {
    super(props);

    this.state = { isHovered: false, showHover: true };

    this._handleMouseEnter = this._handleMouseEnter.bind(this);
    this._handleMouseLeave = this._handleMouseLeave.bind(this);
    this._handleGrant = this._handleGrant.bind(this);
    this._handleRelease = this._handleRelease.bind(this);
  }

  render() {
    const { children } = this.props;
    const child = children(this.state.showHover && this.state.isHovered);

    return cloneElement(Children.only(child), {
      onMouseEnter: this._handleMouseEnter,
      onMouseLeave: this._handleMouseLeave,
      onPressIn: this._handleGrant,
      onPressOut: this._handleRelease,
      onResponderGrant: this._handleGrant,
      onResponderRelease: this._handleRelease,
    });
  }

  _handleMouseEnter() {
    if (isHoverEnabled() && !this.state.isHovered) {
      const { onHoverIn } = this.props;
      if (onHoverIn) {
        onHoverIn();
      }
      this.setState(state => ({ ...state, isHovered: true }));
    }
  }

  _handleMouseLeave() {
    if (this.state.isHovered) {
      const { onHoverOut } = this.props;
      if (onHoverOut) {
        onHoverOut();
      }
      this.setState(state => ({ ...state, isHovered: false }));
    }
  }

  // istanbul ignore next
  _handleGrant() {
    this.setState(state => ({ ...state, showHover: false }));
  }

  // istanbul ignore next
  _handleRelease() {
    this.setState(state => ({ ...state, showHover: true }));
  }
}

Hoverable.propTypes = {
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
  onHoverIn: PropTypes.func,
  onHoverOut: PropTypes.func,
};

export default Hoverable;
