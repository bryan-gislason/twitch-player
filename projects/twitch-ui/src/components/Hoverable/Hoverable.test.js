import { fireEvent, render } from "@twitch-player/testing/dist/unit";
import React from "react";
import { Text, TouchableHighlight } from "react-native";
import Hoverable from "./Hoverable";

jest.mock("../../utils", () => ({
  isHoverEnabled: () => true,
}));

describe("HoverableItem", () => {
  let instance;
  let hoverableItem;

  describe("without press listeners", () => {
    beforeEach(() => {
      instance = render(
        <Hoverable>
          {isHover => (
            <TouchableHighlight testID="hoverable-item">
              <Text>{`${isHover ? "HOVER" : "NO HOVER"}`}</Text>
            </TouchableHighlight>
          )}
        </Hoverable>
      );
      hoverableItem = instance.getByTestId("hoverable-item");
    });

    it("renders the item", () => {
      expect(instance.asFragment()).toMatchSnapshot();
    });

    describe("when mouse enters", () => {
      beforeEach(() => {
        fireEvent.mouseEnter(hoverableItem);
      });

      it("renders as hovered", () => {
        expect(instance.asFragment()).toMatchSnapshot();
      });
    });

    describe("when mouse leaves", () => {
      beforeEach(() => {
        fireEvent.mouseEnter(hoverableItem);
        fireEvent.mouseLeave(hoverableItem);
      });

      it("renders as not hovered", () => {
        expect(instance.asFragment()).toMatchSnapshot();
      });
    });
  });

  describe("with press listeners", () => {
    let spyHoverIn;
    let spyHoverOut;

    beforeEach(() => {
      spyHoverIn = jest.fn();
      spyHoverOut = jest.fn();

      instance = render(
        <Hoverable onHoverIn={spyHoverIn} onHoverOut={spyHoverOut}>
          {isHover => (
            <TouchableHighlight testID="hoverable-item">
              <Text>{`${isHover ? "HOVER" : "NO HOVER"}`}</Text>
            </TouchableHighlight>
          )}
        </Hoverable>
      );
      hoverableItem = instance.getByTestId("hoverable-item");
    });

    it("renders the item", () => {
      expect(instance.asFragment()).toMatchSnapshot();
    });

    describe("when mouse enters", () => {
      beforeEach(() => {
        fireEvent.mouseEnter(hoverableItem);
      });

      it("renders as hovered", () => {
        expect(instance.asFragment()).toMatchSnapshot();
      });

      it("invokes `onHoverIn`", () => {
        expect(spyHoverIn).toHaveBeenCalled();
      });
    });

    describe("when mouse enters after it had already entered", () => {
      beforeEach(() => {
        fireEvent.mouseEnter(hoverableItem);
        fireEvent.mouseEnter(hoverableItem);
      });

      it("invokes `onHoverIn` once", () => {
        expect(spyHoverIn).toHaveBeenCalledTimes(1);
      });
    });

    describe("when mouse leaves", () => {
      beforeEach(() => {
        fireEvent.mouseEnter(hoverableItem);
        fireEvent.mouseLeave(hoverableItem);
      });

      it("renders as not hovered", () => {
        expect(instance.asFragment()).toMatchSnapshot();
      });

      it("invokes `onHoverOut`", () => {
        expect(spyHoverOut).toHaveBeenCalled();
      });
    });

    describe("when mouse leaves after it had already left", () => {
      beforeEach(() => {
        fireEvent.mouseEnter(hoverableItem);
        fireEvent.mouseLeave(hoverableItem);
        fireEvent.mouseLeave(hoverableItem);
      });

      it("invokes `onHoverOut` once", () => {
        expect(spyHoverOut).toHaveBeenCalledTimes(1);
      });
    });
  });
});
