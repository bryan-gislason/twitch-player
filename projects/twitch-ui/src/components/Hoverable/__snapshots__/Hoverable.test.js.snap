// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoverableItem with press listeners renders the item 1`] = `
<DocumentFragment>
  <div
    class="rn-alignItems-1oszu61 rn-backgroundColor-1niwhzg rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    data-focusable="true"
    data-testid="hoverable-item"
    tabindex="0"
  >
    <div
      class="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-1471scf rn-fontFamily-14xgk7a rn-fontSize-1b43r93 rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-fontWeight-gul640 rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
    >
      NO HOVER
    </div>
  </div>
</DocumentFragment>
`;

exports[`HoverableItem with press listeners when mouse enters renders as hovered 1`] = `
<DocumentFragment>
  <div
    class="rn-alignItems-1oszu61 rn-backgroundColor-1niwhzg rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    data-focusable="true"
    data-testid="hoverable-item"
    tabindex="0"
  >
    <div
      class="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-1471scf rn-fontFamily-14xgk7a rn-fontSize-1b43r93 rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-fontWeight-gul640 rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
    >
      HOVER
    </div>
  </div>
</DocumentFragment>
`;

exports[`HoverableItem with press listeners when mouse leaves renders as not hovered 1`] = `
<DocumentFragment>
  <div
    class="rn-alignItems-1oszu61 rn-backgroundColor-1niwhzg rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    data-focusable="true"
    data-testid="hoverable-item"
    tabindex="0"
  >
    <div
      class="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-1471scf rn-fontFamily-14xgk7a rn-fontSize-1b43r93 rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-fontWeight-gul640 rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
    >
      NO HOVER
    </div>
  </div>
</DocumentFragment>
`;

exports[`HoverableItem without press listeners renders the item 1`] = `
<DocumentFragment>
  <div
    class="rn-alignItems-1oszu61 rn-backgroundColor-1niwhzg rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    data-focusable="true"
    data-testid="hoverable-item"
    tabindex="0"
  >
    <div
      class="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-1471scf rn-fontFamily-14xgk7a rn-fontSize-1b43r93 rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-fontWeight-gul640 rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
    >
      NO HOVER
    </div>
  </div>
</DocumentFragment>
`;

exports[`HoverableItem without press listeners when mouse enters renders as hovered 1`] = `
<DocumentFragment>
  <div
    class="rn-alignItems-1oszu61 rn-backgroundColor-1niwhzg rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    data-focusable="true"
    data-testid="hoverable-item"
    tabindex="0"
  >
    <div
      class="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-1471scf rn-fontFamily-14xgk7a rn-fontSize-1b43r93 rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-fontWeight-gul640 rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
    >
      HOVER
    </div>
  </div>
</DocumentFragment>
`;

exports[`HoverableItem without press listeners when mouse leaves renders as not hovered 1`] = `
<DocumentFragment>
  <div
    class="rn-alignItems-1oszu61 rn-backgroundColor-1niwhzg rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    data-focusable="true"
    data-testid="hoverable-item"
    tabindex="0"
  >
    <div
      class="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-1471scf rn-fontFamily-14xgk7a rn-fontSize-1b43r93 rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-fontWeight-gul640 rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
    >
      NO HOVER
    </div>
  </div>
</DocumentFragment>
`;
